### ISSUE TITLE
Receive messages asynchronously from EventHub just as callback

### ISSUE THREAD

**Issue by yunhaoling** [Azure/azure-sdk-for-python#5308]
@raghavan20 commented on [Tue Apr 02 2019](https://github.com/Azure/azure-event-hubs-python/issues/104)

I have looked at the example here (https://github.com/Azure/azure-event-hubs-python/blob/master/examples/recv_async.py#L39), it looks like there is one async iteration to pull data (`receiver.receive`) based on timeout and max batch size.

I wonder is there a simple way to use client and receiver objects to register a callback. The callback would be exercised every time a new message is available in a given IoTHub or EventHub partition. Then I could write my custom business logic in the callback. The client will keep running until there is an error or the application shuts down.

I noticed the class 
```
# ReceiveClientAsync 
async def receive_messages_async(self, on_message_received)
```

but this method is not exercised from sample code I have seen or in `AsyncReceiver`

---

@annatisch commented on [Wed Apr 03 2019](https://github.com/Azure/azure-event-hubs-python/issues/104#issuecomment-479553877)

Thanks @raghavan20 - we currently only support the one means of receiving in the EventHubs SDK.
The method you describe is available in the underlying uAMQP library, which can be used directly with EventHub as seen here:
https://github.com/Azure/azure-uamqp-python/blob/master/samples/test_azure_event_hubs_receive.py#L124

Can I ask why you specifically wish to have your business logic in a callback rather than simply called from your main function?

---

@raghavan20 commented on [Wed Apr 03 2019](https://github.com/Azure/azure-event-hubs-python/issues/104#issuecomment-479744424)

Thanks for your answer. 

> Can I ask why you specifically wish to have your business logic in a callback rather than simply called from your main function?

Sorry, I did not understand your question completely. I am not a Python person so I am not that much familiar with async IO / multithreading libraries.  I thought the whole setup (async) to read from multiple partitions is quite verbose for a library user; And It does not continuously read messages from EventHub according to the example. I thought it would be simpler if we can say which partitions we want to read from and register a callback function. Then the callback function can delegate (queue) incoming messages to another Object in the codebase that can carryout the necessary business logic in its own pace.

In Python SDK for IoTEdge (azure-iothub-device-client), we register a callback function for messages from EventHub (Docker Messaging Bus module).

**Comment by raghavan20** [comment]
When I was referring to azure-iothub-client SDK, I was referring to receiving messages from EdgeHub not EventHub.

It would be nice to register a callback function that takes message and context (such as partition) as method arguments.

**Comment by yunhaoling** [comment]
@raghavan20 Sorry for reaching out to you so late and thanks for the suggestion!

We've recently released the eventhub 5.0.0. We introduced new design and receive callback function can now be registered just as you described.

Please take a try and feel free to provide any kinds of feedback.
Closing this issue for now.
